### PR TITLE
NIAD-860: Create code documentation

### PR DESCRIPTION
**Pull request by n-r-b**
## Description

Updated README documentation for running, developing and deploying the adaptor, covering:
- Configuration
- Debugging
- Checkstyle and spotbugs

Also, made the following changes and updated documentation:
- Removed unused environment variables
- Updated environment variable values for consistency

## Jira Ticket

https://gpitbjss.atlassian.net/browse/NIAD-860

## Checklist

These are items (excluding GitHub Checks) which should be confirmed before a branch is ready to merge.

- [x] Acceptance Criteria met
- [x] Commit messages are meaningful
- [x] Manually tested
- [x] Self-reviewed your code
- [ ] Code reviewed from two other developers
- [x] If your pull request depends on any other, please link them in the description
- [x] main branch is passing/green on CI
- [x] Add/update any relevant documentation

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -27,8 +27,6 @@ services:
       - LAB_RESULTS_MONGO_PASSWORD=
       - LAB_RESULTS_MONGO_OPTIONS=
       - LAB_RESULTS_MONGO_AUTO_INDEX_CREATION=true
-      - LAB_RESULTS_SSL_TRUST_STORE_URL=
```
Comment (n-r-b) [inline docker-compose.yml:30]: Removed as we're not using these

File: docker-compose.yml [hunk 2]
```diff
@@ -27,10 +27,7 @@ services:
...
-      - LAB_RESULTS_SSL_TRUST_STORE_PASSWORD=
       - LAB_RESULTS_MONGO_TTL=P30D
-      - LAB_RESULTS_COSMOS_DB_ENABLED=false
```
Comment (n-r-b) [inline docker-compose.yml:33]: Removed as we're not using these

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/configuration/LabResultsMongoClientConfiguration.java [hunk 3]
```diff
@@ -37,8 +37,6 @@
 
     private Duration ttl;
 
-    private boolean cosmosDbEnabled;
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/configuration/LabResultsMongoClientConfiguration.java:40]: Not used, as far as i can see

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -33,15 +33,14 @@ labresults:
     maxRedeliveries: ${LAB_RESULTS_AMQP_MAX_REDELIVERIES:3}
   mongodb:
     uri: ${LAB_RESULTS_MONGO_URI:mongodb://localhost:27017}
-    database: ${LAB_RESULTS_MONGO_DATABASE_NAME:lab-results}
+    database: ${LAB_RESULTS_MONGO_DATABASE_NAME:labresults}
```
Comment (n-r-b) [inline src/main/resources/application.yml:36]: Renamed for consistency

File: OPERATING.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+within an interchange.
+
+Some log lines are uncorrelated to a specific message. These logs have an empty `<Correlation Id>` value.
+
+Messages published to the adaptor's three AMQP message queues include a CorrelationId header.
+
+# Linking a GP Practice to an Lab Results system
+
+The Lab Results Adaptor and Lab Results system communications synchronise through a sequence number mechanism. Linking a GP 
+Practice to an Lab Results system which have never previously exchanged messages requires no additional setup for
```
Comment (antonlebedjko) [inline OPERATING.md:128]:
```suggestion
Practice to a Lab Results system which have never previously exchanged messages requires no additional setup for 
```

File: OPERATING.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+# Mongodb Database Requirements
+
+* The Lab Results Adaptor and Lab Results system communications synchronise through a sequence number mechanism
+* The Mongodb database preserves this synchronisation
+* Deleting the mongodb database and/or its collections will break the link with the Lab Results system
+
+**Amazon Document DB Tips**
+
+In the "Connectivity & security" tab of the cluster a URI is provided to "Connect to this cluster with an application".
+Replace <username>:<insertYourPasswordHere> with the actual mongo username and password to be used by the application.
```
Comment (andrewwa-kainos) [inline OPERATING.md:32]: `<username>:<insertYourPasswordHere>`
You may need to escape the < and >, or use html entities. This renders just as the colon symbol `:`
```suggestion
Replace \<username\>:\<insertYourPasswordHere\> with the actual mongo username and password to be used by the application.
```

File: OPERATING.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
+# Operating
+
+This document contains requirements and tips for operating the adaptor in a production environment.
+
+# AMQP Message Broker Requirements
+
+* The broker must be configured with a limited number of retries and deadletter queues
```
Comment (andrewwa-kainos) [inline OPERATING.md:7]: I think this should be separate words or hyphenated
`dead letter queues`, or `dead-letter queues`.

File: OPERATING.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+Replace <username>:<insertYourPasswordHere> with the actual mongo username and password to be used by the application.
+The value of `LAB_RESULTS_MONGO_URI` should be set to this value. Since the URI string contains credentials we recommend 
+managing the entire value as a secured secret.
+
+The user must have the `readWrite` role or a custom role with specific privileges.
+
+## Database Collections
+
+The default database name is `labresults` but this can be changed through an environment variable. Each deployment of the
+adaptor MUST have its own database, but multiple database could be hosted by a single cluster. The collection names
```
Comment (andrewwa-kainos) [inline OPERATING.md:41]:
```suggestion
adaptor MUST have its own database, but multiple databases could be hosted by a single cluster. The collection names
```

File: OPERATING.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+### Outbound Sequence Ids
+
+Tracks the sequence numbers used to "link" a GP and to HA using EDIFACT messaging. See 
+"Linking a GP Practice to a Lab Results system" section below for more information.
+
+Collection Name: `outboundSequenceId`
+
+Properties:
+
+* `_id` the key for the sequence in the format <type>-<sender>-<recipient> where:
```
Comment (andrewwa-kainos) [inline OPERATING.md:53]: Need to escape or use entities
```suggestion
* `_id` the key for the sequence in the format \<type\>-\<sender\>-\<recipient\> where:
```

File: OPERATING.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+orchestration solution must be configured to store and aggregate the adaptor's logs.
+
+## Correlation Id
+
+The adaptor uses a correlation id to make tracing individual requests easier. All log lines will contain the text 
+`CorrelationId=<Correlation Id>`. For outbound requests the value of `<Correlation Id>` is either declared in the 
+request headers or randomly generated.
+
+For inbound messages the adaptor generates a correlation id for every message (interchange) downloaded from MESH. 
+An interchange may contain multiple messages, and the adaptor uses same correlation id for every messages
```
Comment (andrewwa-kainos) [inline OPERATING.md:118]:
```suggestion
An interchange may contain multiple messages, and the adaptor uses the same correlation id for all messages 
```

File: OPERATING.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+For inbound messages the adaptor generates a correlation id for every message (interchange) downloaded from MESH. 
+An interchange may contain multiple messages, and the adaptor uses same correlation id for every messages 
+within an interchange.
+
+Some log lines are uncorrelated to a specific message. These logs have an empty `<Correlation Id>` value.
+
+Messages published to the adaptor's three AMQP message queues include a CorrelationId header.
+
+# Linking a GP Practice to an Lab Results system
```
Comment (andrewwa-kainos) [inline OPERATING.md:125]:
```suggestion
# Linking a GP Practice to a Lab Results system
```

File: README.md [hunk 12]
```diff
@@ -15,27 +27,58 @@ Variables without a default value and not marked optional *MUST* be defined for
 | LAB_RESULTS_MESH_OUTBOUND_QUEUE_NAME | lab_results_mesh_outbound | The name of the outbound (to MESH) message queue
 | LAB_RESULTS_MESH_INBOUND_QUEUE_NAME  | lab_results_mesh_inbound  | The name of the inbound (from MESH) message queue
 | LAB_RESULTS_GP_OUTBOUND_QUEUE_NAME   | lab_results_gp_outbound   | The name of the outbound (to GP System) message queue
+| LAB_RESULTS_AMQP_BROKERS             | amqp://localhost:5672     | A comma-separated list of URLs to AMQP brokers (*)
+| LAB_RESULTS_AMQP_USERNAME            |                           | (Optional) username for the AMQP server
+| LAB_RESULTS_AMQP_PASSWORD            |                           | (Optional) password for the AMQP server
+| LAB_RESULTS_AMQP_MAX_REDELIVERIES    | 3                         | The number of times an message will be retried to be delivered to consumer. After exhausting all retires, it will be put on DLQ.<queue_name> dead letter queue
```
Comment (andrewwa-kainos) [inline README.md:33]: Is this actually retries, or total attempts? I.e., does this mean that the maximum number of attempts is 3, or 4, by default?
```suggestion
| LAB_RESULTS_AMQP_MAX_REDELIVERIES    | 3                         | The number of times an message will be retried to be delivered to consumer. After exhausting all retries, it will be put on DLQ.<queue_name> dead letter queue
```

**Review by andrewwa-kainos — CHANGES_REQUESTED** [review]
Also noticed some inconsistency about finishing list items with full-stops or not, just FYI.

File: README.md [hunk 13]
```diff
@@ -93,6 +141,7 @@ The adaptor configuration has sensible defaults for local development. Some over
```
Comment (saileshp-kainos) [inline README.md:141]: Can we add the following lines after `Verify the project structure` to provide an example of the structure?

```suggestion
    Project structure -> SDKs -> add new SDK -> select adoptopenjdk-11.jdk/Contents/Home  (or alternative location)
                      -> Project SDK -> Java 11 (11.0.7)
                      -> Module SDK -> Java 11 (11.0.7)

```


image

File: OPERATING.md (hunk 6 above)
Comment (n-r-b) [reply OPERATING.md:32]: Oh yes, thanks for that 👍

File: OPERATING.md (hunk 7 above)
Comment (n-r-b) [reply OPERATING.md:7]: I'll go with hyphenated.

File: README.md (hunk 13 above)
Comment (n-r-b) [reply README.md:141]: Yes, of course 👍

File: README.md (hunk 12 above)
Comment (n-r-b) [reply README.md]: We've set `LAB_RESULTS_AMQP_MAX_REDELIVERIES` to `3` in the `application.yml` and `docker-compose.yml`.

> jms.redeliveryPolicy.maxRedeliveries controls when an incoming message is rejected based on the number of times it has been redelivered, the default value is (-1) disabled. A value of zero would indicate no message redeliveries are accepted, a value of five would allow a message to be redelivered five times, etc.

https://qpid.apache.org/releases/qpid-jms-0.56.0/docs/index.html

We could re-word this?

Comment (andrewwa-kainos) [reply README.md]: So explicitly retries. I've seen confusion about the distinction between attempts and retries in other projects. We could just add similar comments about -1 and 0.

Also just noticed "an message"

**Review by andrewwa-kainos — DISMISSED** [review]

File: README.md (hunk 12 above)
Comment (n-r-b) [reply README.md]: I've added a comment mentioning the default JMS value

File: OPERATING.md [hunk 14]
```diff
@@ -0,0 +1,151 @@
+# Operating
+
+This document contains requirements and tips for operating the adaptor in a production environment.
+
+# AMQP Message Broker Requirements
+
+* The broker must be configured with a limited number of retries and dead-letter queues
+* It is the responsibility of the GP supplier to configure adequate monitoring against the dead-letter queues that allows ALL undeliverable messages to be investigated fully.
```
Comment (matthewdoreilly) [inline OPERATING.md:8]: Being very picky, but remove the trailing full stop here (consistency with the other list items)

File: OPERATING.md [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+* The broker must use persistent queues to avoid loss of data
+* The GP System must persist the relevant message data before acknowledging the message from the queue to avoid loss of data
+
+**Using AmazonMQ**
+
+* A persistent broker (not in-memory) must be used to avoid data loss.
+* A configuration profile that includes settings for [retry and dead-lettering](https://activemq.apache.org/message-redelivery-and-dlq-handling.html) must be applied.
+* AmazonMQ uses the scheme `amqp+ssl://` but this **MUST** be changed the to `amqps://` when configuring the adaptor.
```
Comment (matthewdoreilly) [inline OPERATING.md:16]: Should be `be changed to`

File: OPERATING.md [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+# AMQP Message Broker Requirements
+
+* The broker must be configured with a limited number of retries and dead-letter queues
+* It is the responsibility of the GP supplier to configure adequate monitoring against the dead-letter queues that allows ALL undeliverable messages to be investigated fully.
+* The broker must use persistent queues to avoid loss of data
+* The GP System must persist the relevant message data before acknowledging the message from the queue to avoid loss of data
+
+**Using AmazonMQ**
+
+* A persistent broker (not in-memory) must be used to avoid data loss.
```
Comment (matthewdoreilly) [inline OPERATING.md:14]: Remove trailing full stops in this list here

File: OPERATING.md [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+* A configuration profile that includes settings for [retry and dead-lettering](https://activemq.apache.org/message-redelivery-and-dlq-handling.html) must be applied.
+* AmazonMQ uses the scheme `amqp+ssl://` but this **MUST** be changed the to `amqps://` when configuring the adaptor.
+
+**Using Azure Service Bus**
+
+* The ASB must use [MaxDeliveryCount and dead-lettering](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-dead-letter-queues#exceeding-maxdeliverycount)
+* Azure Service Bus may require some parameters as part of the URL configuration. For example: `LAB_RESULTS_AMQP_BROKERS=amqps://<NAME>.servicebus.windows.net/;SharedAccessKeyName=<KEY NAME>;SharedAccessKey=<KEY VALUE>`
+
+# Mongodb Database Requirements
```
Comment (matthewdoreilly) [inline OPERATING.md:23]: `MongoDB`

File: README.md [hunk 18]
```diff
@@ -1,12 +1,24 @@
...
-Integration Adaptor to simplify processing of Pathology and Screening results
+Integration Adaptor to simplify processing of Pathology and Screening results.
 
 ## Configuration
 
 The adaptor reads its configuration from environment variables. The following sections describe the environment variables
  used to configure the adaptor. 
  
-Variables without a default value and not marked optional *MUST* be defined for the adaptor to run.
+Variables without a default value and not marked optional are *MUST* be defined for the adaptor to run.
```
Comment (matthewdoreilly) [inline README.md:9]: I'd revert this change. The `are` doesn't read right

File: README.md [hunk 19]
```diff
@@ -15,27 +27,58 @@ Variables without a default value and not marked optional *MUST* be defined for
...
 | LAB_RESULTS_MESH_INBOUND_QUEUE_NAME  | lab_results_mesh_inbound  | The name of the inbound (from MESH) message queue
 | LAB_RESULTS_GP_OUTBOUND_QUEUE_NAME   | lab_results_gp_outbound   | The name of the outbound (to GP System) message queue
+| LAB_RESULTS_AMQP_BROKERS             | amqp://localhost:5672     | A comma-separated list of URLs to AMQP brokers (*)
+| LAB_RESULTS_AMQP_USERNAME            |                           | (Optional) username for the AMQP server
+| LAB_RESULTS_AMQP_PASSWORD            |                           | (Optional) password for the AMQP server
+| LAB_RESULTS_AMQP_MAX_REDELIVERIES    | 3                         | The number of times a message will be retried to be delivered to consumer. After exhausting all retries, it will be put on DLQ.<queue_name> dead-letter queue. The default JMS configuration is (-1) disabled.
+
+(*) Active/Standby: The first broker in the list is always used unless there is an error, in which case the other URLs will be used. At least one URL is required.
+
+### Mongodb Configuration Options
```
Comment (matthewdoreilly) [inline README.md:37]: `MongoDB`.

Sorry to be a pain, but can we update the display name version of MongoDB in these docs? Not a fan of "Mongodb" or "mongodb"

File: README.md [hunk 20]
```diff
@@ -121,25 +172,80 @@ Docker Compose allows running multiple instances behind a nginx load balancer in
...
+
+This command will run all tests (unit & integration) and all static analysis and code style checks. 
+
+The `--continue` flag ensures that all tests and checks will run.
+
+    ./gradlew check --continue
+
 ### Debugging
 
+#### Mongo DB
```
Comment (matthewdoreilly) [inline README.md:226]: `MongoDB`

File: OPERATING.md (hunk 14 above)
Comment (n-r-b) [reply OPERATING.md:8]: I was also thinking about doing this... or adding the full stops for consistency :D

File: README.md (hunk 19 above)
Comment (n-r-b) [reply README.md]: No problem - updated this

File: README.md (hunk 18 above)
Comment (n-r-b) [reply README.md]: done

**Review by saileshp-kainos — APPROVED** [review]
LGTM

**Review by andrewwa-kainos — APPROVED** [review]
